beamer/chain.py

```diff
@@ -183,97 +184,85 @@ def _process_events(self) -> None:
+            claim_request(request, context)
 
-            request = self._context.requests.get(claim.request_id)
-            # As per definition an invalid or expired request cannot be claimed
-            # This gives us a chronological order. The agent should never garbage collect
-            # a request which has active claims
-            assert request is not None, "Active claim for non-existent request"
+        elif request.is_withdrawn:
+            active_claims = any(claim.request_id == request.id for claim in context.claims)
+            if not active_claims:
+                log.debug("Removing withdrawn request", request=request)
+                to_remove.append(request.id)
 
-            # check if claim is an honest claim. Honest claims can be ignored.
-            # This only counts for claims, where the agent is not the filler
-            if claim.valid_claim_for_request(request) and request.filler != self._context.address:
-                claim.ignore()
-                continue
+    for request_id in to_remove:
+        context.requests.remove(request_id)
 
-            if claim.transaction_pending:
-                continue
 
-            if latest_timestamp >= claim.termination:
-                withdraw(claim, self._context)
+def process_claims(context: Context) -> None:
+    log.info("Processing claims", num_claims=len(context.claims))
 
-            if claim.is_claimer_winning or claim.is_challenger_winning:
-                maybe_challenge(claim, self._context)
+    to_remove = []
+    for claim in context.claims:
+        log.debug("Processing claim", claim=claim)
 
-        for claim_id in to_remove:
-            self._context.claims.remove(claim_id)
+        if claim.is_withdrawn:
+            log.debug("Removing withdrawn claim", claim=claim)
+            to_remove.append(claim.id)
+            continue
 
+        request = context.requests.get(claim.request_id)
+        # As per definition an invalid or expired request cannot be claimed
+        # This gives us a chronological order. The agent should never garbage collect
+        # a request which has active claims
+        assert request is not None, "Active claim for non-existent request"
 
-class _TransactionFailed(Exception):
-    def __repr__(self) -> str:
-        return "_TransactionFailed(%r)" % self.__cause__
+        # check if claim is an honest claim. Honest claims can be ignored.
+        # This only counts for claims, where the agent is not the filler
+        if claim.valid_claim_for_request(request) and request.filler != context.address:
+            claim.ignore()
+            continue
 
-    def __str__(self) -> str:
-        return "transaction failed: %s" % self.cause()
+        if claim.transaction_pending:
+            continue
 
-    def cause(self) -> str:
-        return str(self.__cause__)
+        block = context.request_manager.web3.eth.get_block("latest")
+        if claim.termination <= block["timestamp"]:
```

Comment:
It was `latest_timestamp >= claim.termination` before and now `claim.termination <= block["timestamp"]`. So I think to match this you need to remove the `=`.

Reply:
👁️